src/EventListener/AddContentLanguageHeaderListener.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\HttpCache\EventListener;
+
+use eZ\Publish\API\Repository\Values\Content\Content;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+
+final class AddContentLanguageHeaderListener implements EventSubscriberInterface
+{
+    public const CONTENT_LANGUAGE_HEADER = 'xlang';
+
+    /** @var bool */
+    private $isTranslationAware;
+
+    public function __construct(bool $isTranslationAware)
+    {
+        $this->isTranslationAware = $isTranslationAware;
+    }
+
+    public function onKernelResponse(ResponseEvent $event)
+    {
+        if (!$this->isTranslationAware || HttpKernelInterface::MASTER_REQUEST != $event->getRequestType()) {
+            return;
+        }
+
+        $request = $event->getRequest();
+        /** @var \eZ\Publish\API\Repository\Values\Content\Content $content */
+        $content = $request->attributes->get('content');
+        if ($content && $content instanceof Content) {
+            $event->getResponse()->headers->add([self::CONTENT_LANGUAGE_HEADER => $content->getDefaultLanguageCode()]);
```

Comment:
What concerns me here is association with Content language rather than Site language. Have you tested the case where Content item's default language is different than Site language? In that case what is visible on the front-end SiteAccess would be different from what Content Version serves as its default translation.

Reply:
The response is tagged with language code that Content was displayed. For example:
If you have 3 languages: `eng-GB` (main), `pol-PL` and `ger-DE` and SiteAccess configured like this:
```
        site:
            languages: [pol-PL, ger-DE]
```

When Content with only german translation will be accessed in `site` SA, a German translation will be used, as Polish one is not available, and response will be tagged with `ger-DE` language code. Due to that, when the same German translation will be modified, the HTTP cache for the content will be purged. 
To avoid situations, when (using the same example) a German translation is firstly created, then the response is cached and a new Polish translation is created and the cache is not stale, the whole content cache is cleared when the published version is a new translation. This also resolves potential issues with Always Available content.
In addition, when Main Translation is published, and Content is not fully translatable, all other translations will be purged.

Reply:
> The response is tagged with language code that Content was displayed. For example: If you have 3 languages: `eng-GB` (main), `pol-PL` and `ger-DE` and SiteAccess configured like this:
> 
> ```
>         site:
>             languages: [pol-PL, ger-DE]
> ```
> 
> When Content with only german translation will be accessed in `site` SA, a German translation will be used, as Polish one is not available, and response will be tagged with `ger-DE` language code. Due to that, when the same German translation will be modified, the HTTP cache for the content will be purged. To avoid situations, when (using the same example) a German translation is firstly created, then the response is cached and a new Polish translation is created and the cache is not stale, the whole content cache is cleared when the published version is a new translation. This also resolves potential issues with Always Available content. In addition, when Main Translation is published, and Content is not fully translatable, all other translations will be purged.

Fair enough, probably prioritizedFieldLanguageCode is pre-populated with information coming from SiteAccessAware layer.
+1